[VisionGlass] Update phone UI according to connection state

felipeerias opened the pull request:
This PR introduces a ModelView instance which can keep track of the connection state, so that the UI will be updated automatically.

The connection states are:

- DISCONNECTED: the VisionGlass is not plugged to the phone
- CONNECTING: the glasses are plugged and set up has started
- REQUESTING_PERMISSIONS: Wolvic is waiting for the user to grant the required USB permission
- CONNECTED: permissions have been granted and the app is setting up the graphics system
- ACTIVE: everything is ready and the user can use Wolvic
- PERMISSIONS_UNAVAILABLE: error state, the user failed to grant the required permission
- DISPLAY_UNAVAILABLE: error state, the initialization of the graphics system has failed

felipeerias commented:
Built on top of https://github.com/Igalia/wolvic/pull/1281

svillar commented on app/src/main/res/values/strings.xml:
```diff
@@ -2535,4 +2535,7 @@ the Select` button. When clicked it closes all the previously selected tabs -->
     <string name="tray_status_phone">Phone: %1$s</string>
 
+    <!-- Instruct the user to connect their Vision Glass -->
+    <string name="vision_glass_disconnected_message">Please connect Vision Glass</string>
```
I guess we can move this to vision glass' strings.xml ?

felipeerias replied:
Done.

svillar commented on app/src/visionglass/java/com/igalia/wolvic/PlatformActivity.java:
```diff
@@ -155,23 +159,33 @@ protected void onCreate(Bundle savedInstanceState) {
 
+    private void onConnectionStateUpdated(PhoneUIViewModel.ConnectionState connectionState) {
+        Log.d(LOGTAG, "Connection state updated: " + connectionState);
+    }
```
Since you're only using it for logging I think we can convert this into a lambda in the call site.

felipeerias replied:
Done.

svillar commented on app/src/visionglass/java/com/igalia/wolvic/PlatformActivity.java:
```diff
@@ -227,19 +242,28 @@ private void initVisionGlass() {
                 // Note that the system will ask for permissions by itself. This is just a fallback.
                 if (mUSBPermissionRequestCount++ > 1) {
                     showAlertDialog(getString(R.string.phone_ui_usb_alert_dialog_description));
+                    mViewModel.updateConnectionState(PhoneUIViewModel.ConnectionState.PERMISSIONS_UNAVAILABLE);
```
Note that I implemented `showAlertDialog`in a way that it calls `finish()` when the user clicks on the OK button, so I guess we need to remove that behaviour as your patch would in theory allow us not to close the app and wait for the user to unplug/plug the glasses, wouldn't it?

svillar reviewed (changes requested):
I've tried it and got a neverending progress spinner. Steps
1. Start Wolvic with no glasses attached
2. Got the "connect vision glass"
3. Connect the glasses
4. Dialog to allow vision glass is shown. Accept it

Then never ending loop

svillar commented:
I had other issues. Don't remember the exact steps but after starting Wolvic and then connecting the glasses I got the system dialogs and accepted them both. Them I got not one but two Wolvic dialogs reporting errors and then the Wolvic phone UI.

We should be careful because there are tons of potential combinations that the current code takes into account. We should not regress that.

felipeerias commented:
@svillar  I have updated the PR. Here's a short summary of the changes:

- adjusted the UI state transitions
- reset the internal state when the user disconnects the glasses
- after we switch to 3D mode, the system often removes the current display and adds a different one; we used to display an alert in this case but I have removed it
- changing the value of `mPresentationDisplay` and `mActivePresentation` happens at the same time in the `showPresentation()` method
- add a message in the UI when there is an error, instructing the user to try again
- a few other changes, usually having to do with state transitions or different events

svillar reviewed (approved):
This looks great now, and covers most of the cases. There is just one case that is not working fine but I am not sure it could be properly handled as we don't have control over it AFAIK. I'll describe the procedure in a different issue. Let's land this as the improvement is noticeable.
